File: 0912. Sort an Array/0912.java
import java.util.*;

class Solution {
    private void heapify(int[] nums, int n, int i) {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < n && nums[left] > nums[largest]) {
            largest = left;
        }

        if (right < n && nums[right] > nums[largest]) {
            largest = right;
        }

        if (largest != i) {
            int temp = nums[i];
            nums[i] = nums[largest];
            nums[largest] = temp;

            heapify(nums, n, largest);
        }
    }

    private void heapSort(int[] nums) {
        int n = nums.length;

        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(nums, n, i);
        }

        for (int i = n - 1; i > 0; i--) {
            int temp = nums[0];
            nums[0] = nums[i];
            nums[i] = temp;

            heapify(nums, i, 0);
        }
    }

    public int[] sortArray(int[] nums) {
        heapSort(nums);

        return nums;
    }

    public static void main(String[] args) {
        int[] nums = {5, 2, 3, 1};

        System.out.println("Input: nums = " + Arrays.toString(nums));

        System.out.println("Output: " + Arrays.toString(new Solution().sortArray(nums)));
    }
}

File: 0912. Sort an Array/0912.cpp
#include "../streamUtils.hpp"

using namespace std;

class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        heapSort(nums, nums.size());

        return nums;
    }

private:
    void heapify(vector<int>& nums, int n, int i) {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;

        if (l < n && nums[l] > nums[largest])
            largest = l;

        if (r < n && nums[r] > nums[largest])
            largest = r;

        if (largest != i) {
            swap(nums[i], nums[largest]);
            heapify(nums, n, largest);
        }
    }

    void heapSort(vector<int>& nums, int n) {
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(nums, n, i);
        }

        for (int i = n - 1; i > 0; i--) {
            swap(nums[0], nums[i]);
            heapify(nums, i, 0);
        }
    }
};

int main() {

    vector<int> nums = {5,2,3,1};
    
    printInput(getName(nums), nums);

    printOutput(Solution().sortArray(nums));

    return 0;
}
